cumulus/pallets/weight-reclaim-tx/src/tests.rs

```diff
@@ -0,0 +1,480 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg(test)]
+
+use super::*;
+use cumulus_primitives_proof_size_hostfunction::PROOF_RECORDING_DISABLED;
+use frame_support::{assert_ok, derive_impl, pallet_prelude::DispatchClass};
+use sp_runtime::{
+	generic,
+	traits::{BlakeTwo256, DispatchTransaction},
+	BuildStorage,
+};
+use sp_trie::proof_size_extension::ProofSizeExt;
+
+pub type Tx = crate::StorageWeightReclaim<Test, frame_system::CheckWeight<Test>>;
+type AccountId = u64;
+type Extrinsic = generic::UncheckedExtrinsic<AccountId, RuntimeCall, (), Tx>;
+type Block = generic::Block<generic::Header<AccountId, BlakeTwo256>, Extrinsic>;
+
+#[frame_support::runtime]
+mod runtime {
+	#[runtime::runtime]
+	#[runtime::derive(
+		RuntimeCall,
+		RuntimeEvent,
+		RuntimeError,
+		RuntimeOrigin,
+		RuntimeFreezeReason,
+		RuntimeHoldReason,
+		RuntimeSlashReason,
+		RuntimeLockId,
+		RuntimeTask
+	)]
+	pub struct Test;
+
+	#[runtime::pallet_index(0)]
+	pub type System = frame_system::Pallet<Test>;
+
+	#[runtime::pallet_index(1)]
+	pub type WeightReclaimTx = crate::Pallet<Test>;
+}
+
+#[derive_impl(frame_system::config_preludes::TestDefaultConfig)]
+impl frame_system::Config for Test {
```

Comment:
It probably needs a custom `MockWeights: frame_system::ExtensionsWeightInfo` that you can control and assert in tests for `CheckWeight`'s `post_dispatch`. To make the tests more accurate, some mock weights can be used for this pallet as well, instead of `()`.

Reply:
I set up configuration for better tests in https://github.com/paritytech/polkadot-sdk/pull/5234/commits/eb0cdff21f2fc1268e2c6f82dffae4ad98c7dc4c but I will wait for the deccumulating/refining post dispatch info implementation.